Added `defects4j env' subcommand

bkushigian opened the pull request:
# The `defects4j env` subcommand
This commit adds the `env` subcommand to defects4j, which prints the execution environment of `defects4j` for easier debugging.

## Use Case
Suppose your script sets various environment components (`JAVA_HOME`, etc) and then calls  several intermediary scripts that in turn invoke `defects4j`. If `defects4j` doesn't work as expected it can be tough to debug. One of the first things you should do is to ensure that `defects4j` is running in correctly configured environment. This script prints environment info to stdout.

## What it does
`defects4j env` prints various environment information that affect D4J's execution to `stdout`.
For instance, running `defects4j env` on my machine produces:
```
$ framework/bin/defects4j env
PWD.........................../Users/benku/Projects/defects4j
SHELL........................./bin/zsh
TZ............................<UNINITIALIZED>
JAVA_HOME...................../Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
Java Exec...................../usr/bin/java
Java Exec Resolved............/System/Library/Frameworks/JavaVM.framework/Versions/A/Commands/java
Java Version:
  openjdk version "1.8.0_232"
  OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
  OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)
Git version...................git version 2.21.0 (Apple Git-122)
SVN version...................1.10.4
Perl version..................v5.18.4
```

There is probably still some work to be done but I'm PRing early so that I can get feedback.

## Possible Todos
1. [ ] **Add write-to-file option?** Sometimes `defects4j` is called as a subprocess and has stdout redirected (say to `/dev/null`). 
2. [ ] **Add write-to-stderr option?** Similar to 1.
3. [ ] **Other environment vars?** What else could affect a d4j run?

Greg4cr commented:
I think the idea is great. 

>> Add write-to-file option?
I think you should do this

>> Other environment vars?
Nothing immediately comes to mind.

bkushigian commented:
Great. Also, what if we want to call this through an opaque call-stack? say `A` calls `B` calls `defects4j`, and we can't tweak `B`. This is in some sense when we'd want the `env` functionality the most. Is there a good way to handle this? Maybe set `export D4J_PRINT_ENV=true` or something?

rjust commented:
@bkushigian Should we rename this command to print-env?

Also, rather than adding `D4J_PRINT_ENV`, I suggest to call print-env whenever [D4J_DEBUG](https://github.com/rjust/defects4j/blob/master/framework/core/Constants.pm#L51) is true.

bkushigian commented:
@rjust these both sound good, I'll put them in the queue

bkushigian commented:
`defects4j print-env` is the new command. This is toggled by default when running `D4J_DEBUG=1 defects4j...`. Not sure if the `$DEBUG` checking is in the best place or not, or if I should print a header/footer/blank line to distinguish from other output. Currently looks like:
```bash
~/Projects/defects4j$ D4J_DEBUG=1 framework/bin/defects4j info -p Chart
PWD.........................../Users/benku/Projects/defects4j
SHELL........................./bin/zsh
TZ............................America/Los_Angeles
JAVA_HOME...................../Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
Java Exec...................../usr/bin/java
Java Exec Resolved............/System/Library/Frameworks/JavaVM.framework/Versions/A/Commands/java
Java Version:
  openjdk version "1.8.0_232"
  OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
  OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)
Git version...................git version 2.26.2
SVN version...................1.10.4
Perl version..................v5.18.4
Summary of configuration for Project: Chart
--------------------------------------------------------------------------------
    Script dir: /Users/benku/Projects/defects4j/framework
      Base dir: /Users/benku/Projects/defects4j
    Major root: /Users/benku/Projects/defects4j/major
      Repo dir: /Users/benku/Projects/defects4j/project_repos
--------------------------------------------------------------------------------
    Project ID: Chart
       Program: jfreechart
    Build file: /Users/benku/Projects/defects4j/framework/projects/Chart/Chart.build.xml
--------------------------------------------------------------------------------
           Vcs: Vcs::Svn
    Repository: file:///Users/benku/Projects/defects4j/project_repos/jfreechart/trunk
     Commit db: /Users/benku/Projects/defects4j/framework/projects/Chart/commit-db
Number of bugs: 26
--------------------------------------------------------------------------------
```

bkushigian commented:
@rjust two questions
1. my linter is giving an error for `$ARG_ERROR`: `Global Symbol "$ARG_ERROR" requires explicit package name"`. However, it seems to work fine w/out importing `Constants`
2. Using the skew or doesn't work, but using normal or does. With `//`:
JAVA_HOME...................../Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
Use of uninitialized value within %ENV in string at /Users/benku/Projects/defects4j/framework/bin/d4j/d4j-print-env line 95.
JJAVA_HOME....................
Java Exec...................../usr/bin/java
```
With `||`:
```
JAVA_HOME...................../Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
Use of uninitialized value within %ENV in string at /Users/benku/Projects/defects4j/framework/bin/d4j/d4j-print-env line 95.
JJAVA_HOME....................(none)
Java Exec...................../usr/bin/java
```
my $val = "$ENV{$key}" || "(none");
```
However, this prints a warning
```
Use of uninitialized value within %ENV .....
```

bkushigian commented:
Made the following chagnes:
1. Removed `($;$)` syntax
2. Used `my ($a, $b, ...) = @_;` syntax for args
3. Used `//` and unquoted ENV
4. Inlined `indent_lines`
5. Moved `d4j-print-env` to `d4j-env` after conversation w/ Rene

I'm still using shift in `print_multiline_entry` since I want to loop over all but the first param---better way?

@rjust does this look good? The most recent commit changes the file name, so for atomic deltas peruse older commits

bkushigian commented:
@rjust should be ready to go modulo merge conflict---these are simple to fix, just delete `<<<<<<<<`, `============`, and `>>>>>>>>>>>>>` lines. I'd do this myself but it wanted to me 'commit merge' or something, and I didn't know if that would update master directly.
